blog/content/post/2021-02-21-tigerlake-osf.md

```diff
@@ -0,0 +1,183 @@
+available open source firmware options. coreboot development for Tiger Lake
+begun some time ago so that when FSP is released, the build target for Tiger
+Lake RVP should be ready. This however is different from EDK2 MinPlatform. The
+open board implementation is released some time after FSP. But let's start form
+the beginning.
+
+## coreboot
+
+Tiger Lake implementation in coreboot was the first we have tried before going
+with MinPlatform because the latter wasn't available yet. It is quite simple to
+configure the build, since almost all the options are in place when selecting
+the TigerLakeRVP platform. However, one may miss the microcode binary when
+building coreboot. Typically one extract the microcode blob from the original
+BIOS binary shipped with the platform if the microcode is not disclosed or
+publicly available. When extracted simply change the configuration options to
+include microcode external binaries and point to the path with extracted
+microcode, done.
+
+I will omit the build process here since I would like focus more on problems
+and their possible solutions. If you are interested in building coreboot for
+RVP platform please refer to [Booting coreboot on Intel Comet Lake S RVP8 blog post](https://blog.3mdeb.com/2020/2020-08-31-booting-coreboot-on-cometlake/)
+
+The first problem I have encountered is that the platform did not print any
+output on any of the serial consoles, although it is capable of printing it it
+according to the schematics. Thankfully the RVP platforms have 7-segment
+displays for post codes which makes it easy to debug instructions after reset
+vector. It occurred that it stops in the cache as RAM setup. At this point
+there is not much one can do without support or bug history. So I have  sent an
+email to coreboot mailing list and got a reply that older microcode revisions
+had problem with new CAR setup:
+
+https://mail.coreboot.org/hyperkitty/list/coreboot@coreboot.org/thread/7YHWASZX3CQ5U3L7D5CVJCDZSMRNCCXK/
+
+I have followed the advise and turned off the INTEL_CAR_NEM_ENHANCED. I did the
+trick and I could see the serial output on the console. "Now it is a piece of
+cake" I thought... Then I saw the FSP memory init returned an error... Great,
+what now?
+
+Second issue is the memory initialization and configuration. After
+investigating the source code I noted the Tiger Lake RVP mainboard has LPDDR4
+configuration. But wait, my platform has 2 DDR4 SODIMMs... That explains the
+error. The typical different between LPDDR and DDR memory is that the former
+requires an exact mapping of memory signal from CPU to DRAM to be passed to
+FSP. So simply zeroing this configuration should be enough. Also the memory
+initialization was called by `meminit_lpddr4x` which was not suited for DDR4,
+so I have to change it to `meminit_ddr4` with appropriate parameters for SMBus
+SPD addresses. After these modifications I could past through the memory
+initialization.
```

Comment:
@miczyg1 when you plan to provide that contribution? Maybe worth adding

Reply:
I had a private gerrit change with the changes, but it seems to be lost. I recall some changes to private patches in gerrit  were to occur according to information on mailing list. Fortunately I have a local git tree with a copy of the modifications on my PC in the office. Will try to contribute.

Reply:
@miczyg1 at least please push that to Dasharo repo.

Reply:
Added a sentence with a link at the end of coreboot section